packages/flutter_tools/lib/src/commands/apk.dart

```diff
@@ -179,227 +191,342 @@ class ApkCommand extends FlutterCommand {
+
+  if (keystoreInfo == null) {
+    printError('Signing the APK using the debug keystore');
+    keystore = components.debugKeystore;
+    keystorePassword = _kDebugKeystorePassword;
+    keyAlias = _kDebugKeystoreKeyAlias;
+    keyPassword = _kDebugKeystorePassword;
+  } else {
+    keystore = new File(keystoreInfo.keystore);
+    keystorePassword = keystoreInfo.password;
+    keyAlias = keystoreInfo.keyAlias;
+    if (keystorePassword.isEmpty || keyAlias.isEmpty) {
+      printError('Must provide a keystore password and a key alias');
+      return 1;
     }
+    keyPassword = keystoreInfo.keyPassword;
+    if (keyPassword.isEmpty)
+      keyPassword = keystorePassword;
   }
 
-  int _signApk(_ApkBuilder builder, _ApkComponents components, File apk) {
-    File keystore;
-    String keystorePassword;
-    String keyAlias;
-    String keyPassword;
-
-    if (argResults['keystore'].isEmpty) {
-      printError('Signing the APK using the debug keystore');
-      keystore = components.debugKeystore;
-      keystorePassword = _kDebugKeystorePassword;
-      keyAlias = _kDebugKeystoreKeyAlias;
-      keyPassword = _kDebugKeystorePassword;
-    } else {
-      keystore = new File(argResults['keystore']);
-      keystorePassword = argResults['keystore-password'];
-      keyAlias = argResults['keystore-key-alias'];
-      if (keystorePassword.isEmpty || keyAlias.isEmpty) {
-        printError('Must provide a keystore password and a key alias');
-        return 1;
-      }
-      keyPassword = argResults['keystore-key-password'];
-      if (keyPassword.isEmpty)
-        keyPassword = keystorePassword;
+  builder.sign(keystore, keystorePassword, keyAlias, keyPassword, apk);
+
+  return 0;
+}
+
+// Creates a new ApplicationPackage from the Android manifest.
+AndroidApk _getApplicationPackage(String apkPath, String manifest) {
+  if (!FileSystemEntity.isFileSync(manifest))
+    return null;
+  String manifestString = new File(manifest).readAsStringSync();
+  xml.XmlDocument document = xml.parse(manifestString);
+
+  Iterable<xml.XmlElement> manifests = document.findElements('manifest');
+  if (manifests.isEmpty)
+    return null;
+  String id = manifests.toList()[0].getAttribute('package');
+
+  String launchActivity;
+  for (xml.XmlElement category in document.findAllElements('category')) {
+    if (category.getAttribute('android:name') == 'android.intent.category.LAUNCHER') {
+      xml.XmlElement activity = category.parent.parent as xml.XmlElement;
+      String activityName = activity.getAttribute('android:name');
+      launchActivity = "$id/$activityName";
+      break;
     }
+  }
+  if (id == null || launchActivity == null)
+    return null;
+
+  return new AndroidApk(localPath: apkPath, id: id, launchActivity: launchActivity);
+}
 
-    builder.sign(keystore, keystorePassword, keyAlias, keyPassword, apk);
+// Returns true if the apk is out of date and needs to be rebuilt.
+bool _needsRebuild(String apkPath, String manifest) {
```

Comment:
Perhaps also include `.packages` in this check? That will capture dependencies changes, including an engine update.